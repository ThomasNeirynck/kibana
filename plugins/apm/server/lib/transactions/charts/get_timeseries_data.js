import {
  TRANSACTION_DURATION,
  TRANSACTION_RESULT,
  APP_NAME,
  TRANSACTION_TYPE,
  TRANSACTION_NAME
} from '../../../../common/constants';
import moment from 'moment';
import { isNumber, get } from 'lodash';
import { getBucketSize } from '../../helpers/get_bucket_size';

export async function getTimeseriesData({
  appName,
  transactionType,
  transactionName,
  setup
}) {
  const { start, end, client, config } = setup;
  const { intervalString } = getBucketSize(start, end, '1m');

  const params = {
    index: config.get('xpack.apm.indexPattern'),
    body: {
      size: 0,
      query: {
        bool: {
          must: [
            { term: { [APP_NAME]: appName } },
            { term: { [TRANSACTION_TYPE]: transactionType } },
            {
              range: {
                '@timestamp': {
                  gte: start,
                  lte: end,
                  format: 'epoch_millis'
                }
              }
            }
          ]
        }
      },
      aggs: {
        response_times: {
          date_histogram: {
            field: '@timestamp',
            interval: intervalString,
            min_doc_count: 0,
            extended_bounds: {
              min: start,
              max: end
            }
          },
          aggs: {
            avg: {
              avg: { field: TRANSACTION_DURATION }
            },
            pct: {
              percentiles: {
                field: TRANSACTION_DURATION,
                percents: [95, 99]
              }
            }
          }
        },
        overall_avg: {
          avg: { field: TRANSACTION_DURATION }
        },
        rpm_per_status_class: {
          filters: {
            filters: {
              '2xx': {
                range: { [TRANSACTION_RESULT]: { gte: 200, lte: 299 } }
              },
              '3xx': {
                range: { [TRANSACTION_RESULT]: { gte: 300, lte: 399 } }
              },
              '4xx': {
                range: { [TRANSACTION_RESULT]: { gte: 400, lte: 499 } }
              },
              '5xx': { range: { [TRANSACTION_RESULT]: { gte: 500, lte: 599 } } }
            }
          },
          aggs: {
            overall_avg: {
              stats_bucket: {
                buckets_path: 'timeseries>rpm[normalized_value]'
              }
            },
            timeseries: {
              date_histogram: {
                field: '@timestamp',
                interval: intervalString,
                min_doc_count: 0,
                extended_bounds: {
                  min: start.valueOf(),
                  max: end.valueOf()
                }
              },
              aggs: {
                cumsum: {
                  cumulative_sum: {
                    buckets_path: '_count'
                  }
                },
                rpm: {
                  derivative: {
                    buckets_path: 'cumsum',
                    unit: '1m'
                  }
                }
              }
            }
          }
        }
      }
    }
  };

  if (transactionName) {
    params.body.query.bool.must.push({
      term: { [`${TRANSACTION_NAME}.keyword`]: transactionName }
    });
  }

  const resp = await client('search', params);
  const responseTimeBuckets = get(resp, 'aggregations.response_times.buckets');
  const statusBuckets = get(resp, 'aggregations.rpm_per_status_class.buckets');
  const overallAvg = get(resp, 'aggregations.overall_avg.value');

  function getDataFor(key) {
    return get(statusBuckets, `${key}.timeseries.buckets`).map(bucket => {
      return get(bucket, 'rpm.normalized_value', 0) || 0;
    });
  }

  return {
    total_hits: resp.hits.total,
    response_times: responseTimeBuckets.reduce(
      (acc, bucket) => {
        const p95 = bucket.pct.values['95.0'];
        const p99 = bucket.pct.values['99.0'];
        acc.dates.push(moment.utc(bucket.key).toISOString());
        acc.avg.push(bucket.avg.value || 0);
        acc.p95.push((isNumber(p95) && p95) || 0);
        acc.p99.push((isNumber(p99) && p99) || 0);
        return acc;
      },
      { avg: [], p95: [], p99: [], dates: [] }
    ),
    rpm_per_status_class: {
      dates: get(statusBuckets, '2xx.timeseries.buckets').map(bucket => {
        return moment.utc(bucket.key).toISOString();
      }),
      '2xx': getDataFor('2xx'),
      '3xx': getDataFor('3xx'),
      '4xx': getDataFor('4xx'),
      '5xx': getDataFor('5xx')
    },
    rpm_per_status_class_average: {
      '2xx': get(statusBuckets, '2xx.overall_avg.avg', 0) || 0,
      '3xx': get(statusBuckets, '3xx.overall_avg.avg', 0) || 0,
      '4xx': get(statusBuckets, '4xx.overall_avg.avg', 0) || 0,
      '5xx': get(statusBuckets, '5xx.overall_avg.avg', 0) || 0
    },
    weighted_average: overallAvg || 0
  };
}
