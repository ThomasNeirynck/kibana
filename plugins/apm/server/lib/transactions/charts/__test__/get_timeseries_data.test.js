import { getTimeseriesData } from '../get_timeseries_data';
import response from './response.json';

describe('get_timeseries_data', () => {
  let res;
  beforeEach(async () => {
    res = await getTimeseriesData({
      appName: 'myAppName',
      transactionType: 'myTransactionType',
      transactionName: 'myTransactionName',
      setup: {
        start: 1510673413814,
        end: 1510674313814,
        client: jest.fn(() => Promise.resolve(response)),
        config: {
          get: () => 'myIndex'
        }
      }
    });
  });

  it('should match snapshot', () => {
    expect(res).toMatchSnapshot();
  });

  it('should have correct order', () => {
    expect(res.tpm_buckets.map(bucket => bucket.key)).toEqual([
      'HTTP 2xx',
      'HTTP 4xx',
      'HTTP 5xx',
      '3CustomStuff'
    ]);
  });
});
