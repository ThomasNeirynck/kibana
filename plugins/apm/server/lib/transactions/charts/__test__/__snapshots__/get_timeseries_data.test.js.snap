// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get_timeseries_data should match snapshot 1`] = `
Object {
  "dates": Array [
    1510673400000,
    1510673460000,
    1510673520000,
    1510673580000,
    1510673640000,
    1510673700000,
    1510673760000,
    1510673820000,
    1510673880000,
    1510673940000,
    1510674000000,
    1510674060000,
    1510674120000,
    1510674180000,
    1510674240000,
    1510674300000,
  ],
  "response_times": Object {
    "avg": Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      553834.7391304348,
      615674.9459459459,
      464274.07575757575,
      396037.4482758621,
      541597.4912280702,
      775031.17,
      483217.125,
      576777.2307692308,
      438265.46428571426,
    ],
    "p95": Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      2470208.6,
      2910176.0999999996,
      2500612.5,
      2527597.15,
      2468230.1,
      2536518.9,
      2498687.35,
      2637059.1,
      2152816.2,
    ],
    "p99": Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      2534567.92,
      3537133.4499999946,
      2594393.6999999993,
      2601982.08,
      2733127.0500000007,
      2776284.010000006,
      3959511.49,
      4023380.240000001,
      2194837.5700000003,
    ],
  },
  "total_hits": 675,
  "tpm_buckets": Array [
    Object {
      "avg": 63.75,
      "key": "HTTP 2xx",
      "values": Array [
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        68,
        54,
        45,
        97,
        88,
        74,
        65,
        19,
      ],
    },
    Object {
      "avg": 7.375,
      "key": "HTTP 4xx",
      "values": Array [
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        3,
        5,
        7,
        7,
        6,
        10,
        13,
        8,
      ],
    },
    Object {
      "avg": 6,
      "key": "HTTP 5xx",
      "values": Array [
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        3,
        7,
        6,
        10,
        6,
        4,
        0,
        0,
      ],
    },
    Object {
      "avg": 63.75,
      "key": "3CustomStuff",
      "values": Array [
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        68,
        54,
        45,
        97,
        88,
        74,
        65,
        19,
      ],
    },
  ],
  "weighted_average": 557652.1466666666,
}
`;
